app/serialize/jsonld_serializer.py

```diff
@@ -0,0 +1,176 @@
+from typing import Any, Dict, List, Set
+
+import json
+from io import IOBase
+
+from app.kg.graph import Graph
+from app.kg.id_base import IdBase
+from app.kg.iri import IRI
+from app.kg.literal import Literal, PropertyValue
+from app.serialize.graph_serializer import GraphSerializer
+from app.serialize.jsonld_configuration import JsonLDConfiguration
+
+
+class JsonLDSerialializer(GraphSerializer):
+    config: JsonLDConfiguration
+
+    def __init__(self, config: JsonLDConfiguration):
+        self.config = config
+
+    def write(self, out: IOBase, graph: Graph) -> None:
+        results: List[Dict[str, Any]] = []
+        all_node_ids: Set[IRI] = graph.get_ids()
+        aggregate_ids: Set[IRI] = set()
+        default_context = self.config.contexts.get(
+            JsonLDConfiguration.DEFAULT_CONTEXT_IRI, dict()
+        )
+        for node_id in sorted(list(all_node_ids)):
+            meta_properties = graph.get_node_meta_properties(node_id)
+            rdf_type = meta_properties.get(Graph.RDF_TYPE_IRI)
+            if rdf_type and rdf_type in self.config.aggregates:
+                aggregate_ids.add(node_id)
+
+        queue = sorted(list(aggregate_ids))
+        self.write_nodes(
+            results, graph, queue, all_node_ids, aggregate_ids, default_context
+        )
+
+        queue = sorted(list(set(all_node_ids) - aggregate_ids))
+        self.write_nodes(
+            results, graph, queue, all_node_ids, aggregate_ids, default_context
+        )
+
+        outcome: Dict[str, Any] = {}
+        if len(default_context) > 0:
+            outcome["@context"] = default_context
+        outcome["@graph"] = results
+        out.write(json.dumps(outcome))
+        out.flush()
+
+    def write_nodes(
+        self,
+        out_results: List[Dict[str, Any]],
+        graph: Graph,
+        queue: List[IRI],
+        all_node_ids: Set[IRI],
+        aggregate_ids: Set[IRI],
+        default_context: Dict[str, Any],
+    ) -> None:
+        while len(queue) > 0:
+            node_id, queue = queue[0], queue[1:]
```

Comment:
It looks like a usual list walk. Using "for/in" would be easier to read and a bit more efficient.

Reply:
fixed